presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Range;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.JoinNode.Type;
+import io.prestosql.sql.planner.plan.LimitNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.extractCardinality;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.FULL;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.LEFT;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.RIGHT;
+import static io.prestosql.sql.planner.plan.Patterns.Join.type;
+import static io.prestosql.sql.planner.plan.Patterns.join;
+import static io.prestosql.sql.planner.plan.Patterns.limit;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+
+public class PushLimitThroughOuterJoin
+        implements Rule<LimitNode>
+{
+    private static final Capture<JoinNode> CHILD = newCapture();
+
+    private static final Pattern<LimitNode> PATTERN =
+            limit()
+                    .with(source().matching(
+                            join()
+                                    .with(type().matching(type -> isLeftOuter(type) || isRightOuter(type)))
+                                    .capturedAs(CHILD)));
+
+    @Override
+    public Pattern<LimitNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(LimitNode parent, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(CHILD);
+        PlanNode left = joinNode.getLeft();
+        PlanNode right = joinNode.getRight();
+
+        if (isLeftOuter(joinNode.getType()) && !isLimited(left, context.getLookup(), parent.getCount())) {
+            left = new LimitNode(context.getIdAllocator().getNextId(), left, parent.getCount(), false);
```

Comment:
make it partial limit as it relaxes partitioning constraints, see: `io.prestosql.sql.planner.optimizations.AddLocalExchanges.Rewriter#visitLimit`